File: 02-Naive-Bayesian/Python/naive_bayes_scratch.py
import numpy as np

def nbayes_hesapla(H,sorgu):
    n_toplam = H.shape[0]
    if n_toplam <=6:
        raise RuntimeError("VERİ_SETİ_YETERSİZ")
    
    x = H[:,0]
    y = H[:,1]
    sinif = H[:,2]
    n_bir = np.sum( sinif == 1)
    n_eksi = np.sum(sinif == -1)
    
    unique_x = len(np.unique(x))
    unique_y = len(np.unique(y))
    
    p_bir = n_bir / n_toplam
    p_eksi = n_eksi / n_toplam

    def olasilik_Hesapla(sutun, hedef_sinif, sorgu_noktasi, unique_sayisi ):
        n_sinif = np.sum(sinif == hedef_sinif)
        eslesen_adet = np.sum(sutun[(sinif == hedef_sinif)] == sorgu_noktasi)
        return (eslesen_adet+1) / (n_sinif +unique_sayisi)
    
    p_x_sorgu_bir = olasilik_Hesapla(x, 1, sorgu[0], unique_x)
    p_y_sorgu_bir = olasilik_Hesapla(y, 1, sorgu[1], unique_y)
    skor_bir = p_bir * p_x_sorgu_bir * p_y_sorgu_bir

    p_x_sorgu_eksi = olasilik_Hesapla(x, -1, sorgu[0], unique_x)
    p_y_sorgu_eksi = olasilik_Hesapla(y, -1, sorgu[1], unique_y)
    skor_eksi = p_eksi * p_x_sorgu_eksi * p_y_sorgu_eksi 

    if skor_bir > skor_eksi:
        return 1,skor_bir   
    else:
        return -1,skor_eksi

    
H = np.array([ 
    [1 ,3  ,1],
    [2 ,0 ,-1],
    [3 ,1  ,1],
    [1 ,2  ,1],
    [3 ,5 ,-1],
    [4 ,2  ,1],
    [2 ,4 ,-1],
    [1 ,1  ,1],
    [3 ,3 ,-1]
])

while True:
    try:
        a = int(input("Sorgunuzun x değerini giriniz: "))
        b = int(input("Sorgunuzun y değerini giriniz: "))
        try:
            tahmin,skor = nbayes_hesapla(H,[a,b])
            print( f"\n Tahmin edilen sınıf={tahmin}. Skor={skor: .6f} ")
            break
        except RuntimeError:
            print("\033[1;31m" + "HATA !!! Veri seti KNN için çok yetersiz! ".center(100,"*") + "\033[0m")
    except ValueError:
        mesaj = " HATA!!! yalnızca tamsayı değerler ile tahmin yapılabilmektedir. Tekrar değer giriniz! ".center(100,"*")
        print("\033[1;31m" + mesaj + "\033[0m")
    else:
        break

File: 02-Naive-Bayesian/Matlab/naive_bayes_scratch.m
H = [ 1 3  1;
      2 0 -1;
      3 1  1;
      1 2  1;
      3 5 -1;
      4 2  1;
      2 4 -1;
      1 1  1;
      3 3 -1];

x = H(:,1);
y = H(:,2);
sinif = H(:,3);

sorgu_x = input("sorgunuzun x değerini giriniz: ");
sorgu_y = input("sorgunuzun y değerini giriniz: ");
sorgu = [sorgu_x,sorgu_y];

n_toplam = size(H,1); %toplam veri sayısını matrisin satır sayısından alıyoruz
n_bir = sum(sinif == 1); %sınıfı 1 olan toplam veri sayısını tutar
n_eksi = sum(sinif == -1); %sınıfı -1 olan toplam veri sayısını tutar

unique_x = length(unique(x)); %x sütunundaki benzersiz değer sayısı
unique_y = length(unique(y)); %x sütunundaki benzersiz değer sayısı

p_bir = n_bir / n_toplam; %veri setindeki bir verinin sınıfının 1 olma olasılığı (prior probability)
p_eksi = n_eksi / n_toplam; %veri setindeki bir verinin sınıfının 1 olma olasılığı (prior probability)

x_bir_adet = sum(x(sinif == 1) == sorgu(1)); %sınıfı 1 olan verilerin x değerleri ile sorgunun x değerini karşılaştır. Sorgunun x'ine eşit olanların toplamı
y_bir_adet = sum(y(sinif == 1) == sorgu(2)); %sınıfı 1 olan verilerin y değerleri ile sorgunun y değerini karşılaştır. Sorgunun y'ine eşit olanların toplamı
                                             %yani sorgudaki x değeri veri setinde kaç kez 1 sınıfına ait olarak görülür
x_eksi_adet = sum(x(sinif == -1) == sorgu(1)); %aynı işlemler -1 sınıfı için de tekrar edilir
y_eksi_adet = sum(y(sinif == -1) == sorgu(2));

p_x_sorgu_bir = (x_bir_adet + 1) / (n_bir + unique_x); %sorgudaki x değeriin veri setinde 1 sınıfına ait olma olasılığı. paylara eklenen +1 laplace smoothingdir 
p_y_sorgu_bir = (y_bir_adet + 1) / (n_bir + unique_y); %sorgudaki y değeriin veri setinde 1 sınıfına ait olma olasılığı 

skor_bir = p_bir * p_x_sorgu_bir * p_y_sorgu_bir; %bulduğumuz sonuçlarla sorgunun bütün olarak 1 sınıfına ait olma olasılığının hesabı (posterior probability)

p_x_sorgu_eksi = (x_eksi_adet + 1) / (n_eksi + unique_x); %sorgudaki x değeriin veri setinde -1 sınıfına ait olma olasılığı 
p_y_sorgu_eksi = (y_eksi_adet + 1) / (n_eksi + unique_y); %sorgudaki y değeriin veri setinde -1 sınıfına ait olma olasılığı 

skor_eksi = p_eksi * p_x_sorgu_eksi * p_y_sorgu_eksi; %bulduğumuz sonuçlarla sorgunun bütün olarak 1 sınıfına ait olma olasılığının hesabı (posterior probability)

if skor_bir > skor_eksi
    fprintf('Tahmin Edilen Sınıf: 1 (Skor: %f)\n', skor_bir); 
else                                                                %olasılığı yüksek olan sınıfa ait olarak yazdırılır
    fprintf('Tahmin Edilen Sınıf: -1 (Skor: %f)\n', skor_eksi);
end
